src/NexusMods.Telemetry/EventSender.cs

```diff
@@ -0,0 +1,253 @@
+        sb.Append(metadata.CurrentTime.Hour);
+        sb.Append("&m="); // The current minute (local time)
+        sb.Append(metadata.CurrentTime.Minute);
+        sb.Append("&s="); // The current second (local time)
+        sb.Append(metadata.CurrentTime.Second);
+
+        // NOTE(erri120): maybe something for later
+        // sb.Append("&res=1920x1080"); // The resolution of the device the visitor is using
+
+        sb.Append("\"");
+        sb.CopyTo(writer);
+    }
+
+    private static async ValueTask SendEvents(ReadOnlyMemory<byte> data, HttpClient httpClient, CancellationToken cancellationToken)
+    {
+        using var request = new HttpRequestMessage(HttpMethod.Post, Constants.MatomoTrackingEndpoint);
+        request.Content = new ReadOnlyMemoryContent(data);
+        request.Content.Headers.ContentType = new MediaTypeHeaderValue(MediaTypeNames.Application.Json);
+
+        var response = await httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
+        response.EnsureSuccessStatusCode();
+    }
+
+    /// <summary>
+    /// Copies the queued events into a secondary array and sorts them by id.
+    /// </summary>
+    private void PrepareArrays()
+    {
+        _insertRingBuffer.CopyTo(_sortedReadingCopy, index: 0);
+        Array.Sort(_sortedReadingCopy, static (a, b) => a.Item3.CompareTo(b.Item3));
+    }
+
+    /// <summary>
+    /// Creates a slice of the array to exclude already seen events.
+    /// </summary>
+    private static ReadOnlySpan<ValueTuple<EventDefinition, EventMetadata, ulong>> PrepareSpan(ValueTuple<EventDefinition, EventMetadata, ulong>[] input, ulong highestSeenId)
+    {
+        Debug.Assert(input.Length == Limit);
+
+        var sliceStartIndex = 0;
+        for (var i = 0; i < Limit; i++)
+        {
+            var id = input[i].Item3;
+            if (id > highestSeenId) break;
+            sliceStartIndex += 1;
+        }
+
+        if (sliceStartIndex >= Limit) return ReadOnlySpan<ValueTuple<EventDefinition, EventMetadata, ulong>>.Empty;
+        return input.AsSpan(start: sliceStartIndex);
+    }
+
+    /// <summary>
+    /// Creates a user-agent than Matomo understands.
+    /// </summary>
+    internal static string CreateUserAgent(OSPlatform platform, OperatingSystem osVersion)
+    {
+        // https://lx3rzx16x9.csb.app/#Arch%20Linux%206.12
+        if (platform == OSPlatform.Linux)
+        {
+            return $"Arch Linux {osVersion.Version.ToString(fieldCount: 2)}";
+        }
+
+        // https://lx3rzx16x9.csb.app/#Windows%20NT%2010.0
+        if (platform == OSPlatform.Windows)
+        {
+            return $"Windows NT {osVersion.Version.ToString(fieldCount: 2)}";
+        }
+
+        // https://lx3rzx16x9.csb.app/#Macintosh;%20Intel%20Mac%20OS%20X%2010_1
+        if (platform == OSPlatform.OSX)
+        {
+            return $"Macintosh; Intel Mac OS X {osVersion.Version.Major}_{osVersion.Version.Minor}";
```

Comment:
We can't detect the CPU architecture? For distinguishing Apple silicon?

Reply:
You can ignore this method, it's not used. Matomo user agent detection is really whack.